src/Swift.Runtime/src/Swift/SwiftString.cs

```diff
@@ -124,15 +130,28 @@ public SwiftString(string str)
         {
             fixed (byte* utf8BytesPtr = utf8Bytes)
             {
-                _payload = PInvoke_Create(utf8BytesPtr, utf8Bytes.Length, 1);
+                _buffer = PInvoke_Create(utf8BytesPtr, utf8Bytes.Length, 1);
+                fixed (void* buffer = &_buffer)
+                    _payload = new SwiftHandle((IntPtr)buffer);
             }
         }
     }
 
     /// <summary>
     /// Gets the length of string.
     /// </summary>
-    public int Length => (int)PInvoke_GetLength(_payload);
+    public int Length
+    {
+        get
+        {
+            bool _success = false;
+            _payload.DangerousAddRef(ref _success);
```

Comment:
Curious - these are here so that if some other thread calls Dispose while the PInvoke is on the stack, that we don't destroy the value, right?
Technically this should be wrapped in try/finally, so that if the pinvoke throws, we still release the handle. For example: https://github.com/dotnet/runtime/blob/be6d8d01a848a1de0674e51c6d243b65404f2e0f/src/libraries/Common/src/Interop/Unix/System.Native/Interop.Poll.cs#L36

Reply:
I wonder if this is necessary - does it really protect all cases of races between one thread calling dispose while other is performing an operation?

Reply:
If called on an instance method or passed as an argument, the Dispose won't release the native handle until it is released at the end of the call. If Dispose is invoked before the call, the `DangerousAddRef` will throw `ObjectDisposedException`.

Do you have in mind any other cases?

Reply:
Thanks I understand that case - I was wondering if there are other ways to make it wrong. I guess technically the unsafe accessor to the payload can be used to make this wrong. I remember we had a plan to remove it (that it is there for MarshalToSwift or something similar and that once we change that we won't need to expose the raw payload outside of the projected type).

Reply:
Do you mean Swift native handle by unsafe accessor? MarshalToSwift is used only for generic types; otherwise, we pass a SafeHandle if the payload is opaque, or a Buffer if the payload is enregistered. In the future, it's possible we always operate on a copy of the native handle and encapsulating the payload.